### PR TITLE
[WIP] dubbogo compass metrics Prometheus support

### PR DESCRIPTION
**Pull request by flixgithub**
**What this PR does**:
Prometheus for dubbogo first step, more metrics collect please stay tuned
1. Compass metrics collect summary implement

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: monitor/dubbogo_exporter.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monitor
+
+import (
```
Comment (AlexStocks) [inline monitor/dubbogo_exporter.go]: split

File: monitor/collector/collector.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
+	for _, m := range e.metrics {
+		ch <- m
+	}
+}
+
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
```
Comment (AlexStocks) [inline monitor/collector/collector.go]: the lock scope is so large, sir. in my opinion, u just need to use the read lock in the for-loop.

File: metrics/histogram.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+type Histogram interface {
+	Metric
+	Sampling
+	Counting
+	/**
+	 * Adds a recorded value.
+	 *
+	 * @param value the length of the value
+	 */
+	Update(value int)
```
Comment (flycash) [inline metrics/histogram.go]: What's the difference between Update and UpdateN. I think that Update = UpdateN(1). Right?

Comment (flycash) [inline metrics/histogram.go]: What's the difference between Update and UpdateN. I think that Update = UpdateN(1). Right?

Comment (AlexStocks) [reply metrics/histogram.go]: @flycash  Agree with u. just one func is enough.

File: monitor/collector/collector.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package collector
+
+import (
+	"github.com/apache/dubbo-go/metrics/impl"
+	"github.com/prometheus/client_golang/prometheus"
+	"sync"
```
Comment (zouyx) [inline monitor/collector/collector.go]: split?

File: metrics/impl/histogram_impl.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+ */
+
+package impl
+
+import (
+	"github.com/apache/dubbo-go/metrics"
+	"time"
+)
+
+type HistorgramImpl struct {
```
Comment (zouyx) [inline metrics/impl/histogram_impl.go]: i am not agree to use HistorgramImpl ,it seems not very intuitive
my suggest:```PrometheusHistorgram``` or ```HistorgramManager```

File: monitor/dubbogo_exporter.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"net/http"
+)
+
+const (
+	namespace = "dubbogo" // For Prometheus metrics.
+)
+
+var (
+	listenAddr       = flag.String("web.listen-port", "9001", "An port to listen on for web interface and telemetry.")
```
Comment (hxmhlt) [inline monitor/dubbogo_exporter.go]: It is necessary to open the way to configure prometheus port and namespace.

**Comment by fangyincheng** [comment]
don't use `impl` directory. refer to `filter` directory.

**Comment by AlexStocks** [comment]
@flixgithub pls review those comments and improve them.

**Pull request merged**